src/reflect/scala/reflect/internal/util/Collections.scala

```diff
@@ -61,6 +61,19 @@ trait Collections {
     head
   }
 
+  final def sameElementsEquals(thiss: List[AnyRef], that: List[AnyRef]): Boolean = {
+    // Probably immutable, so check reference identity first (it's quick anyway)
+    (thiss eq that) || {
+      var these = thiss
+      var those = that
+      while (!these.isEmpty && !those.isEmpty && these.head.equals(those.head)) {
```

Comment:
I've got a bit of an aversion to `nonEmpty` (a mixed in trait method) in hot code because in some cases it is harder for JIT to devirtualize and inline. Not sure whether that's the case here.